src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.Utf8JsonStreamReader.cs

```diff
@@ -0,0 +1,1801 @@
+                        {
+                            try
+                            {
+#pragma warning disable CS0612 // Type or member is obsolete
+                                packageSpec.HasVersionSnapshot = PackageSpecUtility.IsSnapshotVersion(version);
+#pragma warning restore CS0612 // Type or member is obsolete
+                                packageSpec.Version = PackageSpecUtility.SpecifySnapshot(version, snapshotValue);
+                            }
+                            catch (Exception ex)
+                            {
+                                throw FileFormatException.Create(ex, version, packageSpec.FilePath);
+                            }
+                        }
+                    }
+                    else
+                    {
+                        jsonReader.Skip();
+                    }
+                }
+            }
+            packageSpec.Name = name;
+            packageSpec.FilePath = name == null ? null : Path.GetFullPath(packageSpecPath);
+
+#pragma warning disable CS0612 // Type or member is obsolete
+            if (!wasPackOptionsSet)
+            {
+                packageSpec.Owners = [];
+                packageSpec.PackOptions = new PackOptions()
+                {
+                    PackageType = Array.Empty<PackageType>()
+                };
+                packageSpec.Tags = [];
+            }
+
+            if (isMappingsNull)
+            {
+                packageSpec.PackOptions.Mappings = null;
+            }
+#pragma warning restore CS0612 // Type or member is obsolete
+
+            packageSpec.RuntimeGraph = new RuntimeGraph(
+                runtimeDescriptions ?? Enumerable.Empty<RuntimeDescription>(),
+                compatibilityProfiles ?? Enumerable.Empty<CompatibilityProfile>());
+
+            packageSpec.Name ??= packageSpec.RestoreMetadata?.ProjectName;
+
+            // Use the project.json path if one is set, otherwise use the project path
+            packageSpec.FilePath ??= packageSpec.RestoreMetadata?.ProjectJsonPath
+                    ?? packageSpec.RestoreMetadata?.ProjectPath;
+
+            return packageSpec;
+        }
+
+        internal static void ReadCentralTransitiveDependencyGroup(
+            ref Utf8JsonStreamReader jsonReader,
+            IList<LibraryDependency> results,
+            string packageSpecPath)
+        {
+            if (jsonReader.Read() && jsonReader.TokenType == JsonTokenType.StartObject)
+            {
+                while (jsonReader.Read() && jsonReader.TokenType == JsonTokenType.PropertyName)
+                {
+                    var propertyName = jsonReader.GetString();
+                    if (string.IsNullOrEmpty(propertyName))
+                    {
+                        throw FileFormatException.Create(
+                            "Unable to resolve dependency ''.",
+                            packageSpecPath);
+                    }
+
+                    if (jsonReader.Read())
```

Comment:
There is no `else` for this statement. My understanding is that `Read` only returns false at the end of the file, in which case it means this JSON file is incomplete. Traditionally, parsers would throw an exception at the point of syntax error (like missing value following a property name), but this code appears to return an incomplete value.

Reply:
So you're right that read should only return false at the end of the file. That said Read() will throw an error if we're at the end of the file at this point. The file needs to be structured to be able to call read on the reader and be at a StartObject. I added a test to confirm this. 

Utf8JsonReader throws an exception if we try to read and hit the end of file when it's several layers deep into an object already.

Reply:
I see. So from a practical point of view, `Read()` is never expected to return false, except for potentially the last read of the root object's deserialization (but from another PR, you found that Newtosoft.Json doesn't even validate that end of an object is also the end of the file).

This makes all the `if (jsonReader.Read())` conditions unnecessary nesting.

I'm wondering from an API design point of view, if `Utf8JsonStreamReader`'s `Read()` method should be changed to return `void`. We can add a `EnsureEndOfFile()` if we want to check that there is no contents at the end of the root object.

On the other hand, I feel guilty when my comments feel like scope creep, so this probably really doesn't matter, as `Utf8JsonStreamReader` is internal, and nuget should write the files that are read, so we have very high confidence there will not be any errors in the json.